# diff for OpenMage 20.15.0
diff --git a/errors/processor.php b/errors/processor.php
index 9bee05b380..bec9aa06fb 100644
--- a/errors/processor.php
+++ b/errors/processor.php
@@ -368,7 +368,7 @@ class Error_Processor
     /**
      * @throws Exception
      */
-    public function saveReport(array $reportData)
+    public function saveReport(array $reportData, $t = null)
     {
         $this->reportData = $reportData;
         $this->reportId   = abs((int) (microtime(true) * random_int(100, 1000)));
@@ -398,6 +398,22 @@ class Error_Processor
         }
         $this->_setReportUrl();
 
+        // https://github.com/luigifab/openmage-sentry
+        // @see Mage_Core_Model_App::_initCurrentStore()
+        try {
+            global $sentry;
+            if ($sentry && is_object($sentry)) {
+                $text = implode("\n", $reportData);
+                if (empty($t))
+                    $sentry->captureMessage('Report: '.$text, 'fatal', ['source' => 'errors:report', 'report' => $this->reportId]);
+                else
+                    $sentry->captureException($t,
+                        'Report '.$this->reportId.' catched by Processor->saveReport()'."\n".$text,
+                        ['source' => 'errors:report', 'report' => $this->reportId]);
+            }
+        }
+        catch (Throwable $ts) { }
+
         if (headers_sent()) {
             echo '<script type="text/javascript">';
             echo "window.location.href = encodeURI('{$this->reportUrl}');";
diff --git a/errors/report.php b/errors/report.php
index 6e73080bec..2fdfa6619d 100644
--- a/errors/report.php
+++ b/errors/report.php
@@ -26,7 +26,7 @@ require_once 'processor.php';
 $processor = new Error_Processor();
 
 if (isset($reportData) && is_array($reportData)) {
-    $processor->saveReport($reportData);
+    $processor->saveReport($reportData, $e ?? null);
 }
 
 $processor->processReport();
