# multiparts diff for OpenMage 20.15.0
# ##########################################################################################################################################
# this is OpenMage PR 3596
diff --git a/app/code/core/Mage/Core/Block/Abstract.php b/app/code/core/Mage/Core/Block/Abstract.php
index f33b14a031e..7d25687b626 100644
--- a/app/code/core/Mage/Core/Block/Abstract.php
+++ b/app/code/core/Mage/Core/Block/Abstract.php
@@ -183,6 +183,13 @@
      */
     protected $_app;
 
+    /**
+     * To not render the same block multiple times with getChildHtml()
+     *
+     * @var bool
+     */
+    protected $_isRendered = false;
+
     /**
      * Initialize factory instance
      */
@@ -580,7 +587,9 @@
             }
             $out = '';
             foreach ($children as $child) {
-                $out .= $this->_getChildHtml($child->getBlockAlias(), $useCache);
+                if (!$child->isRendered()) {
+                    $out .= $this->_getChildHtml($child->getBlockAlias(), $useCache);
+                }
             }
             return $out;
         } else {
@@ -914,6 +923,8 @@
         if (Mage::getStoreConfig('advanced/modules_disable_output/' . $this->getModuleName())) {
             return '';
         }
+
+        $this->_isRendered = true;
         $html = $this->_loadCache();
         if ($html === false) {
             $translate = Mage::getSingleton('core/translate');
@@ -1591,4 +1601,14 @@
 
         return Mage::helper($helperAlias)->isModuleOutputEnabled($moduleName);
     }
+
+    /**
+     * Flag
+     *
+     * @return bool
+     */
+    public function isRendered()
+    {
+        return $this->_isRendered;
+    }
 }
diff --git a/app/design/adminhtml/default/default/layout/main.xml b/app/design/adminhtml/default/default/layout/main.xml
index b2a2ef42283..be57ffdc0cd 100644
--- a/app/design/adminhtml/default/default/layout/main.xml
+++ b/app/design/adminhtml/default/default/layout/main.xml
@@ -89,6 +89,7 @@
                 <action method="addItem"><type>js</type><name>mage/adminhtml/wysiwyg/tinymce/setup.js</name><params/><if/><condition>can_load_tiny_mce</condition></action>
 
                 <block type="core/html_calendar" name="head.calendar" as="calendar" template="page/js/calendar.phtml"/>
+                <block type="adminhtml/text_list" name="before_cssjs"/>
             </block>
 
             <block type="adminhtml/page_notices" name="global_notices" as="global_notices" template="page/notices.phtml" />
diff --git a/app/design/adminhtml/default/default/template/page/head.phtml b/app/design/adminhtml/default/default/template/page/head.phtml
index 741efe119b9..2d92574d2e4 100644
--- a/app/design/adminhtml/default/default/template/page/head.phtml
+++ b/app/design/adminhtml/default/default/template/page/head.phtml
@@ -20,11 +20,12 @@
     var BASE_URL = '<?php echo $this->getUrl('*') ?>';
     var SKIN_URL = '<?php echo $this->getSkinUrl() ?>';
     var FORM_KEY = '<?php echo $this->getFormKey() ?>';
-    <?php # BC: cast to INT in case of non-existing method getLoadingTimeout() in 3rd-party code?>
+    <?php // BC: cast to INT in case of non-existing method getLoadingTimeout() in 3rd-party code ?>
     var LOADING_TIMEOUT = <?php echo (int)$this->getLoadingTimeout() ?>;
 //]]>
 </script>
 
+<?php echo $this->getChildHtml('before_cssjs') ?>
 <?php echo $this->getCssJsHtml() ?>
 
 <?php if ($this->getCanLoadExtJs()): ?>
@@ -50,4 +51,4 @@
 </script>
 
 <?php echo $this->helper('core/js')->getTranslatorScript() ?>
-<?php echo $this->getChildHtml(); ?>
+<?php echo $this->getChildHtml() ?>
diff --git a/app/design/frontend/base/default/layout/page.xml b/app/design/frontend/base/default/layout/page.xml
index 82da2714fc3..82b79528a7b 100644
--- a/app/design/frontend/base/default/layout/page.xml
+++ b/app/design/frontend/base/default/layout/page.xml
@@ -37,12 +37,12 @@
                 <action method="addJs"><script>varien/menu.js</script></action>
                 <action method="addJs"><script>mage/translate.js</script></action>
                 <action method="addJs"><script>mage/cookies.js</script></action>
-
-                <block type="page/js_cookie" name="js_cookies" template="page/js/cookie.phtml"/>
-
                 <action method="addCss"><stylesheet>css/styles.css</stylesheet></action>
                 <action method="addCss"><stylesheet>css/widgets.css</stylesheet></action>
                 <action method="addCss"><stylesheet>css/print.css</stylesheet><params>media="print"</params></action>
+
+                <block type="page/js_cookie" name="js_cookies" template="page/js/cookie.phtml"/>
+                <block type="core/text_list" name="before_cssjs"/>
             </block>
 
             <block type="core/text_list" name="after_body_start" as="after_body_start" translate="label">
diff --git a/app/design/frontend/base/default/template/page/html/head.phtml b/app/design/frontend/base/default/template/page/html/head.phtml
index e8b95fc628b..7c09e1e722b 100644
--- a/app/design/frontend/base/default/template/page/html/head.phtml
+++ b/app/design/frontend/base/default/template/page/html/head.phtml
@@ -17,6 +17,7 @@
 <meta name="keywords" content="<?php echo htmlspecialchars($this->getKeywords()) ?>" />
 <meta name="robots" content="<?php echo htmlspecialchars($this->getRobots()) ?>" />
 <link rel="icon" href="<?php echo $this->getFaviconFile(); ?>" type="image/x-icon" />
+<?php echo $this->getChildHtml('before_cssjs') ?>
 <?php echo $this->getCssJsHtml() ?>
 <?php echo $this->getChildHtml() ?>
 <?php echo $this->helper('core/js')->getTranslatorScript() ?>
# ##########################################################################################################################################
# this is OpenMage PR 3679
diff --git a/app/Mage.php b/app/Mage.php
index dee273c2d3e..a478f86cb56 100644
--- a/app/Mage.php
+++ b/app/Mage.php
@@ -968,21 +968,61 @@
     public static function printException(Throwable $e, $extra = '')
     {
         if (self::$_isDeveloperMode) {
-            print '<pre>';
-
+            if (PHP_SAPI != 'cli') {
+                print '<pre style="white-space:pre-wrap;">';
+            }
             if (!empty($extra)) {
                 print $extra . "\n\n";
             }
-
-            print $e->getMessage() . "\n\n";
-            print $e->getTraceAsString();
-            print '</pre>';
+            print get_class($e) . ': ' . $e->getMessage() . "\n\n";
+            if (PHP_SAPI != 'cli') {
+                $file = $e->getFile();
+                $line = $e->getLine();
+                if (!empty($_SERVER['MAGE_VSCODE_LINKS']) || !empty($_ENV['MAGE_VSCODE_LINKS'])) {
+                    // @see https://code.visualstudio.com/docs/editor/command-line#_opening-vs-code-with-urls
+                    print preg_replace_callback('#(\#\d+ )([^(]+)\((\d+)\): #', static function ($data) {
+                        return
+                            $data[1] .
+                            '<a href="vscode://file/' . $data[2] . ':' . $data[3] . '">' . $data[2] . '</a>' .
+                            '(' . $data[3] . '): ';
+                    }, $e->getTraceAsString()) . "\n";
+                    print '  thrown in <a href="vscode://file/' . $file . ':' . $line . '"><b>' . $file . '</b></a>';
+                } elseif (!empty($_SERVER['MAGE_PHPSTORM_LINKS']) || !empty($_ENV['MAGE_PHPSTORM_LINKS'])) {
+                    // @todo phpstorm doc
+                    print preg_replace_callback('#(\#\d+ )([^(]+)\((\d+)\): #', static function ($data) {
+                        return
+                            $data[1] .
+                            '<a href="phpstorm://open?url=file:/' . $data[2] . '&line=' . $data[3] . '">' . $data[2] . '</a>' .
+                            '(' . $data[3] . '): ';
+                    }, $e->getTraceAsString()) . "\n";
+                    print '  thrown in <a href="phpstorm://open?url=file:/' . $file . '&line=' . $line . '"><b>' . $file . '</b></a>';
+                } elseif (!empty($_SERVER['MAGE_OPENFILEEDITOR_LINKS']) || !empty($_ENV['MAGE_OPENFILEEDITOR_LINKS'])) {
+                    // @see https://github.com/luigifab/webext-openfileeditor
+                    print preg_replace_callback('#(\#\d+ )([^(]+)\((\d+)\): #', static function ($data) {
+                        return
+                            $data[1] .
+                            '<span class="openfileeditor" data-line="' . $data[3] . '">' . $data[2] . '</span>' .
+                            '(' . $data[3] . '): ';
+                    }, $e->getTraceAsString()) . "\n";
+                    print '  thrown in <span class="openfileeditor" data-line="' . $line . '"><b>' . $file . '</b></span>';
+                } else {
+                    print $e->getTraceAsString() . "\n";
+                    print '  thrown in <b>' . $file . '</b>';
+                }
+                print ' on line <b>' . $line . '</b>' . "\n";
+            } else {
+                print $e->getTraceAsString() . "\n";
+                print '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine() . "\n";
+            }
+            if (PHP_SAPI != 'cli') {
+                print '</pre>';
+            }
         } else {
             $reportData = [
-                (!empty($extra) ? $extra . "\n\n" : '') . $e->getMessage(),
-                $e->getTraceAsString(),
+                (empty($extra) ? '' : $extra . "\n\n") . get_class($e) . ': ' . $e->getMessage(),
+                $e->getTraceAsString() . "\n" .
+                    '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine(),
             ];
-
             // retrieve server data
             if (isset($_SERVER['REQUEST_URI'])) {
                 $reportData['url'] = $_SERVER['REQUEST_URI'];
@@ -990,7 +1030,6 @@
             if (isset($_SERVER['SCRIPT_NAME'])) {
                 $reportData['script_name'] = $_SERVER['SCRIPT_NAME'];
             }
-
             // attempt to specify store as a skin
             try {
                 $storeCode = self::app()->getStore()->getCode();
diff --git a/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php b/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
index 5ad0ad18ad8..06fe7c70af8 100644
--- a/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
+++ b/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
@@ -33,6 +33,7 @@
      * @param int $errorCode
      * @param string $errorMessage
      * @param string $errorFile
+     * @param int $errLine
      * @return bool
      */
     public function handlePhpError($errorCode, $errorMessage, $errorFile, $errLine)
diff --git a/app/code/core/Mage/Core/Block/Template.php b/app/code/core/Mage/Core/Block/Template.php
index 2598f3304a1..18725c4c0f8 100644
--- a/app/code/core/Mage/Core/Block/Template.php
+++ b/app/code/core/Mage/Core/Block/Template.php
@@ -245,15 +245,50 @@
 <div style="position:relative; border:1px dotted {$cacheHintStatusColor}; margin:6px 2px; padding:18px 2px 2px 2px; zoom:1;">
 <div style="position:absolute; left:0; top:0; padding:2px 5px; background:{$cacheHintStatusColor}; color:white; font:normal 11px Arial;
 text-align:left !important; z-index:998;text-transform: none;" onmouseover="this.style.zIndex='999'"
-onmouseout="this.style.zIndex='998'" title="{$fileName}">{$fileName}</div>
+onmouseout="this.style.zIndex='998'" title="{$fileName}">
 HTML;
+            if (!empty($_SERVER['MAGE_VSCODE_LINKS']) || !empty($_ENV['MAGE_VSCODE_LINKS'])) {
+                // @see https://code.visualstudio.com/docs/editor/command-line#_opening-vs-code-with-urls
+                echo '<a href="vscode://file/' . $this->_viewDir . DS . $fileName . '">' . $fileName . '</a></div>';
+            } elseif (!empty($_SERVER['MAGE_PHPSTORM_LINKS']) || !empty($_ENV['MAGE_PHPSTORM_LINKS'])) {
+                // @todo phpstorm doc
+                echo '<a href="phpstorm://open?url=file:/' . $this->_viewDir . DS . $fileName . '">' . $fileName . '</a></div>';
+            } elseif (!empty($_SERVER['MAGE_OPENFILEEDITOR_LINKS']) || !empty($_ENV['MAGE_OPENFILEEDITOR_LINKS'])) {
+                // @see https://github.com/luigifab/webext-openfileeditor
+                echo '<span class="openfileeditor" data-file="' . $this->_viewDir . DS . $fileName . '">' . $fileName . '</span></div>';
+            } else {
+                echo $fileName . '</div>';
+            }
+
             if (Mage::app()->getStore()->isAdmin() ? self::$_showTemplateHintsBlocksAdmin : self::$_showTemplateHintsBlocks) {
                 $thisClass = static::class;
                 echo <<<HTML
 <div style="position:absolute; right:0; top:0; padding:2px 5px; background:{$cacheHintStatusColor}; color:blue; font:normal 11px Arial;
 text-align:left !important; z-index:998;text-transform: none;" onmouseover="this.style.zIndex='999'" onmouseout="this.style.zIndex='998'"
-title="{$thisClass}">{$thisClass}</div>
+title="{$thisClass}">
 HTML;
+                try {
+                    if (!empty($_SERVER['MAGE_VSCODE_LINKS']) || !empty($_ENV['MAGE_VSCODE_LINKS'])) {
+                        $reflector = new ReflectionClass($thisClass);
+                        $fileClass = $reflector->getFileName();
+                        // @see https://code.visualstudio.com/docs/editor/command-line#_opening-vs-code-with-urls
+                        echo '<a href="vscode://file/' . $fileClass . '">' . $thisClass . '</a></div>';
+                    } elseif (!empty($_SERVER['MAGE_PHPSTORM_LINKS']) || !empty($_ENV['MAGE_PHPSTORM_LINKS'])) {
+                        $reflector = new ReflectionClass($thisClass);
+                        $fileClass = $reflector->getFileName();
+                        // @todo phpstorm doc
+                        echo '<a href="phpstorm://open?url=file:/' . $fileClass . '">' . $thisClass . '</a></div>';
+                    } elseif (!empty($_SERVER['MAGE_OPENFILEEDITOR_LINKS']) || !empty($_ENV['MAGE_OPENFILEEDITOR_LINKS'])) {
+                        $reflector = new ReflectionClass($thisClass);
+                        $fileClass = $reflector->getFileName();
+                        // @see https://github.com/luigifab/webext-openfileeditor
+                        echo '<span class="openfileeditor" data-file="' . $fileClass . '">' . $thisClass . '</span></div>';
+                    } else {
+                        echo $thisClass . '</div>';
+                    }
+                } catch (Throwable $t) {
+                    echo $thisClass . '</div>';
+                }
             }
         }
 
diff --git a/app/code/core/Mage/Core/functions.php b/app/code/core/Mage/Core/functions.php
index d643769e7f2..79cbc410642 100644
--- a/app/code/core/Mage/Core/functions.php
+++ b/app/code/core/Mage/Core/functions.php
@@ -145,11 +145,10 @@
         default => $errorMessage .= "Unknown error ($errno)",
     };
 
-    $errorMessage .= ": {$errstr}  in {$errfile} on line {$errline}";
     if (Mage::getIsDeveloperMode()) {
-        throw new Exception($errorMessage);
+        Mage::printException(new ErrorException($errorMessage . ' ' . $errstr, 0, $errno, $errfile, $errline));
     } else {
-        Mage::log($errorMessage, Zend_Log::ERR);
+        Mage::log($errorMessage . ': ' . $errstr . ' in ' . $errfile . ' on line ' . $errline, Zend_Log::ERR);
         return null;
     }
 }
diff --git a/app/code/core/Mage/Cron/Model/Observer.php b/app/code/core/Mage/Cron/Model/Observer.php
index fa1ee0b18c8..8ef37a6bfe5 100644
--- a/app/code/core/Mage/Cron/Model/Observer.php
+++ b/app/code/core/Mage/Cron/Model/Observer.php
@@ -324,7 +324,8 @@
                 ->setFinishedAt(date(Varien_Db_Adapter_Pdo_Mysql::TIMESTAMP_FORMAT));
         } catch (Exception $e) {
             $schedule->setStatus($errorStatus)
-                ->setMessages($e->__toString());
+                ->setMessages(get_class($e) . ': ' . $e->getMessage() . "\n" . $e->getTraceAsString() . "\n" .
+                    '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine());
         }
 
         if ($schedule->getIsError()) {
# ##########################################################################################################################################
# this is for Sentry module
diff --git a/app/Mage.php b/app/Mage.php
index 1f50465e73..65ed72cb2b 100644
--- a/app/Mage.php
+++ b/app/Mage.php
@@ -27,6 +27,7 @@
 
 if (!empty($_SERVER['MAGE_IS_DEVELOPER_MODE']) || !empty($_ENV['MAGE_IS_DEVELOPER_MODE'])) {
     Mage::setIsDeveloperMode(true);
+    error_reporting(E_ALL);
     ini_set('display_errors', '1');
     ini_set('error_prepend_string', '<pre>');
     ini_set('error_append_string', '</pre>');
@@ -930,8 +931,7 @@
 
             $message = addcslashes($message, '<?');
             $loggers[$file]->log($message, $level);
-        } catch (Exception $e) {
-        }
+        } catch (Exception $e) { }
     }
 
     /**
@@ -945,7 +945,16 @@
             return;
         }
         $file = self::getStoreConfig('dev/log/exception_file');
-        self::log("\n" . $e->__toString(), Zend_Log::ERR, $file);
+        self::log(
+            (self::$_isDeveloperMode ? 'isDeveloperMode:on ' : 'isDeveloperMode:off ') .
+                (self::getStoreConfigFlag('dev/sentry/active') ? 'sentry:on' : 'sentry:off') .
+                "\n" . $e->getMessage() . "\n" . $e->getTraceAsString() .
+                "\n" . '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine(),
+            Zend_Log::ERR, $file);
+        if (!self::$_isDeveloperMode) {
+            global $sentry;
+            $sentry->captureException($e, null, ['source' => 'mage:logException']);
+        }
     }
 
     /**
@@ -975,10 +984,17 @@
     /**
      * Display exception
      */
-    public static function printException(Throwable $e, $extra = '')
+    public static function printException(Throwable $e, $extra = '', $source = '')
     {
+        try {
+            while (ob_get_level() > 0) {
+                ob_end_clean();
+            }
+        } catch (Exception $e) { }
+
         if (self::$_isDeveloperMode) {
             if (PHP_SAPI != 'cli') {
+                @http_response_code(500); // @todo this prevent display of ajax grid error
                 print '<pre style="white-space:pre-wrap;">';
             }
             if (!empty($extra)) {
@@ -1024,6 +1040,7 @@
                 print $e->getTraceAsString() . "\n";
                 print '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine() . "\n";
             }
+            print '  catched by Mage::printException()' . $source . ' at ' . date('c') . "\n";
             if (PHP_SAPI != 'cli') {
                 print '</pre>';
             }
@@ -1031,7 +1048,8 @@
             $reportData = [
                 (empty($extra) ? '' : $extra . "\n\n") . get_class($e) . ': ' . $e->getMessage(),
                 $e->getTraceAsString() . "\n" .
-                    '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine(),
+                    '  thrown in ' . $e->getFile() . ' on line ' . $e->getLine() . "\n" .
+                    '  catched by Mage::printException()' . $source . ' at ' . date('c')
             ];
             // retrieve server data
             if (isset($_SERVER['REQUEST_URI'])) {
@@ -1044,8 +1062,7 @@
             try {
                 $storeCode = self::app()->getStore()->getCode();
                 $reportData['skin'] = $storeCode;
-            } catch (Exception $e) {
-            }
+            } catch (Exception $e) { }
 
             require_once(self::getBaseDir() . DS . 'errors' . DS . 'report.php');
         }
diff --git a/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php b/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
index 5bdec0e513..0415b2db42 100644
--- a/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
+++ b/app/code/core/Mage/Api/Model/Server/Handler/Abstract.php
@@ -36,9 +36,11 @@
      * @param int $errLine
      * @return bool
      */
-    public function handlePhpError($errorCode, $errorMessage, $errorFile, $errLine)
+    public function handlePhpError($errorCode, $errorMessage, $errorFile, $errLine, $context = [])
     {
         Mage::log($errorMessage . ' in ' . $errorFile . ' on line ' . $errLine, Zend_Log::ERR);
+        global $sentry;
+        $sentry->handleError($errorCode, $errorMessage, $errorFile, $errLine, $context);
         if (in_array($errorCode, [E_ERROR, E_USER_ERROR, E_RECOVERABLE_ERROR])) {
             $this->_fault('internal');
         }
diff --git a/app/code/core/Mage/Core/Model/App.php b/app/code/core/Mage/Core/Model/App.php
index d44240102c..52c41ab26f 100644
--- a/app/code/core/Mage/Core/Model/App.php
+++ b/app/code/core/Mage/Core/Model/App.php
@@ -507,9 +507,17 @@
             $this->_checkCookieStore($scopeType);
             $this->_checkGetStore($scopeType);
         }
-        $this->_useSessionInUrl = (bool) $this->getStore()->getConfig(
-            Mage_Core_Model_Session_Abstract::XML_PATH_USE_FRONTEND_SID,
-        );
+
+        // @see https://github.com/luigifab/openmage-sentry
+        $store = $this->getStore();
+        global $sentry;
+        if (!$sentry) {
+            $sentry = new Luigifab_Sentry_Model_Client();
+            $sentry->initHandler($store);
+        }
+
+        $this->_useSessionInUrl = (bool) $store->getConfig(Mage_Core_Model_Session_Abstract::XML_PATH_USE_FRONTEND_SID);
+
         return $this;
     }
 
diff --git a/app/code/core/Mage/Core/Model/Email/Template/Filter.php b/app/code/core/Mage/Core/Model/Email/Template/Filter.php
index ea1cc4871e..4db12d177f 100644
--- a/app/code/core/Mage/Core/Model/Email/Template/Filter.php
+++ b/app/code/core/Mage/Core/Model/Email/Template/Filter.php
@@ -549,6 +549,9 @@
         try {
             $value = parent::filter($value);
         } catch (Exception $e) {
+            if (Mage::getIsDeveloperMode()) {
+                throw $e;
+            }
             $value = '';
             Mage::logException($e);
         }
@@ -569,6 +572,9 @@
         try {
             $result = parent::_getVariable($value, $default);
         } catch (Exception $e) {
+            if (Mage::getIsDeveloperMode()) {
+                throw $e;
+            }
             $result = '';
             Mage::logException($e);
         }
diff --git a/app/code/core/Mage/Core/functions.php b/app/code/core/Mage/Core/functions.php
index a9ce749a80..5b59931b5e 100644
--- a/app/code/core/Mage/Core/functions.php
+++ b/app/code/core/Mage/Core/functions.php
@@ -168,7 +168,7 @@
     };
 
     if (Mage::getIsDeveloperMode()) {
-        Mage::printException(new ErrorException($errorMessage . ' ' . $errstr, 0, $errno, $errfile, $errline));
+        Mage::printException(new ErrorException($errorMessage . ' ' . $errstr, 0, $errno, $errfile, $errline), '', ' via mageCoreErrorHandler()');
     } else {
         Mage::log($errorMessage . ': ' . $errstr . ' in ' . $errfile . ' on line ' . $errline, Zend_Log::ERR);
         return null;
